File: 405.py
class Solution:
    def toHex(self, num: int) -> str:
        # time: O(1), space: O(1)
        if num == 0:
            return "0"
        elif num < 0:
            num += 2 ** 32
        h, r = "0123456789abcdef", []
        while num:
            r.append(h[num & 15])
            num //= 16
        return "".join(r[::-1])

File: 405.java
class Solution {
    public String toHex(int num) {
        // time: O(1), space: O(1)
        if (num == 0) return "0";
        StringBuilder str = new StringBuilder();
        char[] h = "0123456789abcdef".toCharArray();
        while (num != 0) {
            int r = num & 15;
            str.append(h[r]);
            num >>>= 4;
        }
        return str.reverse().toString();
    }
}